### PR TITLE
fix(git-remote): drive multi-round fetch as a v0 stateless-RPC client loop (#117)

### LINKED ISSUE #117: git-remote-gitlawb deadlocks on incremental (multi-round) fetch
## Summary

The `git-remote-gitlawb` helper bridges git's stateful upload-pack `connect` stream onto a single stateless HTTP POST, so it deadlocks on an **incremental (multi-round) fetch**. `read_upload_pack_request` treats a flush pkt (`0000`) as "keep reading" and only breaks on the `done\n` pkt-line, and `handle_connect` issues exactly one POST after `done`. In an incremental fetch git sends `want`s, then a `have`-batch terminated by a flush, and **blocks waiting for the server's ACK/NAK before sending `done`**. The helper is blocked reading stdin for a `done` that never comes; git is blocked reading the helper's stdout for a response that is never sent. The fetch hangs until the request timeout.

A fresh clone (the client sends `want`s + `done` in one shot) works, which is why this was not caught earlier.

## Where

- `crates/gitlawb-node/../git-remote-gitlawb/src/main.rs` — `read_upload_pack_request` (breaks only on `done\n` or EOF; `continue`s past a flush) and the single-POST flow in `handle_connect`.

## Reproduced by execution

Live repro against a local node: built a ~50-commit shared history, cloned via `gitlawb://`, advanced the server by a few commits, then `git fetch origin`.

- Fresh clone (control): exit 0, ~1s.
- Incremental fetch, few haves (git short-circuits, sends `want` + haves + `done` in one shot): exit 0, ~0s.
- Incremental fetch, >32 haves (forces multi-round): **hung the full 40s, killed by `timeout` (exit 124)**.

`GIT_TRACE_PACKET` confirmed git sent 1 `want`, 32 `have` lines, a flush `0000`, and **zero `done`**, then blocked. Only the initial ref advertisement ever came back.

## Impact

Any fetch where the client has more than ~32 commits of overlapping history with the server (the common "pull updates into an existing clone") hangs indefinitely. Not a security issue; a functional break in the helper's fetch path.

## Options

1. Implement git's stateless-RPC multi-round negotiation (a POST per have-batch, streaming ACK/NAK back), or
2. advertise/implement the `stateless-connect` capability instead of `connect`.

Pre-existing; surfaced while verifying the receive-pack advertisement fix end to end.

### PR DESCRIPTION
**Pull request by beardthelion**
Fixes #117.

`git-remote-gitlawb` advertised the `connect` capability, so git spoke the stateful native protocol, but `handle_connect` collapsed the exchange into a single GET + single POST. A multi-round fetch (more than ~32 commits of overlapping history, the common "pull into an existing clone") deadlocked: git sent a flush-terminated `have` batch and blocked for an ACK/NAK the helper never sent, while the helper blocked reading for a `done` git never sent.

Phase 2 for `git-upload-pack` is now a per-round v0 smart-HTTP stateless-RPC client loop, the role git's own `remote-curl` plays. `read_upload_pack_round` returns one negotiation round at a time (at a flush, `done`, or EOF) instead of buffering past flushes; `negotiate_upload_pack` captures the wants once and POSTs a self-contained request per flush-terminated batch (wants + one flush + every have accumulated so far + exactly one terminator), streams each ACK/NAK back to git, and stops on the `done` round. The node's `git upload-pack --stateless-rpc` keeps no state between POSTs, so wants and all prior haves are re-sent every round and no intermediate flush survives into a body.

Signing is preserved per round: every POST carries the Phase-1 decision (signed after the 404 escalation for a private repo, anonymous for a public one), and a mid-negotiation denial surfaces through the sanitized error path rather than reading as an empty or successful fetch. The `git-receive-pack` (push) path is unchanged, and no node code is touched.

## Verification

- Unit + mock coverage: the deadlock repro (blocking reader, was RED at the 2s timeout), multi-round accumulation asserted byte-for-byte on the wire, per-round signing for private and public fetches (including the must-not cases), mid-negotiation denial surfacing, single-round non-regression, and the branch-coverage edges. Load-bearing checks confirmed by mutation.
- A committed real-git integration test (`tests/real_git_fetch.rs`) drives a real `git fetch` (forced to >=2 rounds) through the built helper against a real `git upload-pack --stateless-rpc`, so the stateful-to-stateless bridging is executed, not reasoned.
- Verified against a running node: an incremental fetch with a real two-round negotiation exits 0 with the correct object graph (the exact case that hung before), and a private repo fetched anonymously surfaces the sanitized 404 promptly with no hang and no leak.

## Follow-up

The one case this does not make work is an incremental fetch of a repo with a path-scoped withheld subtree: the node's `upload_pack_excluding` sends its pack on the first POST, which real git rejects mid-negotiation. The helper forwards and terminates cleanly, so it is a node-side issue, filed as #191. A committed test here (`real_git_withheld_shaped_first_post`) records that rejection so a node-side fix can verify against it.

STRATEGY track: stabilize.

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved multi-round `git-upload-pack --stateless-rpc` negotiation with correct per-round termination, streamed forwarding, and accumulated “haves/wants”.
  * Prevented hangs and tightened edge-case handling (EOF/malformed streams), including more accurate fetch completion tracking and error surfacing.
  * Ensured fetch request signing behavior remains correct for public vs private fetches across rounds.
* **Tests**
  * Expanded real end-to-end `git fetch` coverage (multi-round boundaries, withheld-first behavior, negotiation denial, signing verification, and protocol edge cases).
  * Strengthened bounded subprocess timeout/cleanup coverage to avoid stdout/stderr deadlocks.
* **Chores**
  * Updated Unix-focused dev dependencies to support the enhanced integration harness.

### CODE REVIEW

**Review by jatmn — CHANGES_REQUESTED** [review]
I found issues that need to be addressed before this is ready.

## Findings

- [P2] Do not convert an aborted negotiation into a `done` request
  `crates/git-remote-gitlawb/src/main.rs:610`
  A nonempty EOF is explicitly documented here as the case where Git aborted mid-negotiation, but it is combined with `RoundEnd::Done`, receives a synthetic `0009done\n`, and is sent to the node.  Cancelling a fetch after writing a partial have batch therefore still issues an authorized (and, for a private fetch, signed) upload-pack request, making the node generate and stream a pack to a client that has gone away.  EOF should terminate the helper without a POST; only the actual `done` pkt should finalize the stateless request.  The new `nonempty_eof_batch_posts_as_final_round` test currently locks in the incorrect behavior.

- [P2] Drain the real Git child output while waiting for it
  `crates/git-remote-gitlawb/tests/real_git_fetch.rs:266`
  `fetch_with_helper` pipes both stdout and stderr, then only polls `try_wait` until the process exits.  A fetch that emits more progress or error output than an OS pipe can hold blocks in Git before it exits, so this harness reaches its 30-second timeout and reports the protocol deadlock even when the helper is making progress.  Drain both streams concurrently while enforcing the deadline (or do not pipe them) so the new load-bearing integration test can distinguish a real negotiation hang from pipe backpressure.

- [P3] Make the integration-test repositories cleanup-safe
  `crates/git-remote-gitlawb/tests/real_git_fetch.rs:356`
  The server and clone directories are deleted only after every assertion succeeds.  Any timeout, failed fetch, `fsck` failure, or panic exits before `cleanup`, leaving real Git repositories in the shared temporary directory; the predictable pid/counter names can also poison a later run if a pid is reused.  Return RAII-owned temporary directories (or use `tempfile`) so cleanup runs while unwinding as well.

**Comment by beardthelion** [comment]
All three addressed on `e3c322c`.

**Aborted EOF no longer POSTs** (F1). The fetch loop folded `RoundEnd::Eof` into the `Done` arm, so a nonempty have-batch ending in EOF (git aborted mid-negotiation) got a synthetic `0009done` and was POSTed, making the node generate and stream a pack to a client that had gone away. Split the arm: only a real `done` pkt finalizes the stateless request; a nonempty EOF terminates without a POST (the accumulated haves are discarded, there is no one to serve). The `nonempty_eof_batch_posts_as_final_round` test that locked in the old behavior is flipped to assert no POST.

**Concurrent drain** (F2). Factored the spawn/deadline logic into `run_bounded`, which now drains stdout and stderr on reader threads while enforcing the deadline, so a fetch emitting more than an OS pipe buffer can no longer block git before it exits and trip the timeout as a false deadlock. Covered by a load-bearing test: a child writing ~140 KiB to both streams completes under the drained runner and deadlocks (deadline trips) under the old try_wait-only one.

**RAII cleanup** (F3). The server and clone now live under a single `tempfile::TempDir` that drops on unwind, so a timeout, failed fetch, `fsck` failure, or panic cleans up while unwinding instead of leaking real repos; the randomized root also removes the predictable pid/counter naming. A `catch_unwind` test asserts the tree is gone after a mid-test panic.

Each fix is RED->GREEN with the guard reverted to confirm it is load-bearing.

**Review by jatmn — CHANGES_REQUESTED** [review]
I found issues that need to be addressed before this is ready.

## Findings

- [P2] Kill the entire fetch process tree before joining the drainers
  `crates/git-remote-gitlawb/tests/real_git_fetch.rs:305`
  On the timeout path this kills only the top-level `git fetch` process and then joins the `read_to_end` threads. Its `git-remote-gitlawb` descendant inherits those stdout/stderr pipe FDs, so a helper stuck in its HTTP request keeps the pipes open after Git dies; the joins then wait for that helper's 300-second request timeout instead of returning the promised 30-second deadlock result. Start the fetch in a process group and terminate/reap that group (or otherwise reap descendants) before joining the readers. This is also the unresolved current-head CodeRabbit request.

- [P2] Make the withheld-path integration test prove it reached the withheld response
  `crates/git-remote-gitlawb/tests/real_git_fetch.rs:518`
  The non-success branch accepts every prompt `git fetch` failure and only prints the observed POST count. A regression that fails before the helper sends the first request to the shim—for example a broken advertisement, helper lookup, or connection—therefore passes as the expected rejected filtered pack, so this new test no longer verifies the path it claims to record. Require at least one POST (and preferably assert the expected protocol failure when a failure remains acceptable) before accepting the nonzero exit.

File: crates/git-remote-gitlawb/tests/real_git_fetch.rs [hunk 1]
```diff
@@ -0,0 +1,538 @@
...
+    let stderr = err_reader.join().expect("stderr reader");
+    (
+        completed,
+        std::process::Output {
+            status,
+            stdout,
+            stderr,
+        },
+    )
+}
```
[the remark this answers is not in the document]
Comment (beardthelion) [reply crates/git-remote-gitlawb/tests/real_git_fetch.rs:605]: > If the remote helper is still blocked in its HTTP timeout when the deadline fires, it can keep the inherited pipe handles open and leave the reader threads hanging forever. Kill the helper process group (or otherwise reap descendants) before joining.

Fixed in 4a3f59c. `run_bounded` now spawns the child with `process_group(0)` and, on the timeout branch, tears down the whole group before joining the readers: SIGTERM, poll `kill(-pgid, 0)` for ESRCH, escalate to SIGKILL after a short grace, bounded cap, then reap. That closes every inherited pipe write-end (the `git fetch` leader and the `git-remote-gitlawb` descendant), so the joins return at the deadline instead of the helper's ~300s HTTP timeout. Mirrors the existing idiom in `gitlawb-node/src/git/smart_http.rs`; `libc` is added as a `cfg(unix)` dev-dep and non-unix keeps the `child.kill()` fallback.

New regression `run_bounded_reaps_descendants_holding_the_pipe`: a backgrounded `sleep 10` inherits the stderr pipe and outlives a 1s-deadline leader. Reverting to the old leader-only `child.kill()` turns it RED (readers block ~10s); with the group teardown it returns in ~1s.

**Comment by beardthelion** [comment]
@jatmn both P2s addressed in 4a3f59c, each verified RED/GREEN by execution.

**F1 (kill the whole process tree before joining):** `run_bounded` now spawns with `process_group(0)` and, on the timeout branch, tears the whole group down (SIGTERM, ESRCH poll on `kill(-pgid, 0)`, SIGKILL escalation after a grace, bounded cap, then reap) before joining the readers. That closes the `git-remote-gitlawb` descendant's inherited pipe write-ends too, so the joins return at the 30s deadline instead of the helper's ~300s HTTP timeout. Mirrors `gitlawb-node/src/git/smart_http.rs`; `libc` added as a `cfg(unix)` dev-dep, non-unix keeps the `child.kill()` fallback. New regression `run_bounded_reaps_descendants_holding_the_pipe`: reverting to the leader-only kill makes it RED (readers block ~10s on a backgrounded descendant holding the pipe), green with the group teardown.

**F2 (prove the withheld test reached the withheld response):** the non-success branch now asserts `posts >= 1` before treating a nonzero exit as the withheld rejection, so a failure before the first POST (broken advertisement, helper lookup, connection) can no longer masquerade as the expected rejection. Forcing a pre-POST failure (fetch against a dead URL, 0 POSTs) makes the new assert fire; the real shim run stays green.

Full crate green (42 unit + 5 integration), fmt and workspace clippy clean.

**Review by jatmn — CHANGES_REQUESTED** [review]
I found issues that need to be addressed before this is ready.

## Findings

- [P2] Keep the large-output regression test portable to the shipped Windows target
  `crates/git-remote-gitlawb/tests/real_git_fetch.rs:558`
  This test unconditionally spawns `sh -c 'seq …'`, but the release workflow ships `git-remote-gitlawb` for `x86_64-pc-windows-msvc` and neither `sh` nor `seq` is a guaranteed Windows dependency.  Consequently `cargo test -p git-remote-gitlawb` fails at `spawn child` on a normal Windows checkout before it tests the drainer.  Gate the shell-based fixture to Unix or replace it with a platform-native/Rust output generator.

- [P2] Terminate the remote-helper tree in the non-Unix timeout path
  `crates/git-remote-gitlawb/tests/real_git_fetch.rs:379`
  The Unix path creates and kills a process group because the `git-remote-gitlawb` descendant inherits Git's stdout/stderr pipe handles; the non-Unix fallback kills only the top-level `git` process and then joins `read_to_end` drainers.  When a timed-out fetch leaves the helper blocked in its 300-second HTTP request, those inherited handles remain open and the joins wait for that timeout rather than returning after the promised 30 seconds.  Use a Windows process-tree mechanism (for example a Job Object or equivalent) before joining the drainers, and cover that path on its supported platform.

**Comment by beardthelion** [comment]
@jatmn both P2s addressed in 34040be, each verified by execution.

**F1 (portable large-output fixture):** the test binary now re-invokes itself as the output generator (an `#[ignore]`d fixture gated on `GL_TEST_FIXTURE=emit` writing ~136 KiB to both streams), so `run_bounded_drains_large_output_without_deadlock` keeps running on windows rather than being cfg-gated away; no `sh`/`seq` remains in the file. The guard stays load-bearing: reverting `run_bounded` to poll-then-read turns it RED (`completed=false`, 11.03s), GREEN with the drain restored.

**F2 (non-unix tree teardown):** the timeout branch now calls `reap_tree`, which runs `taskkill /T /F /PID <leader>` (your "or equivalent": a tree kill shipped in System32, no Job Object dev-dependency or unsafe in test code) before the drainer joins, then reaps the leader, mirroring `reap_group`. The descendants test dropped its `#[cfg(unix)]` and drives a self-exec leader plus pipe-holding grandchild, so the taskkill path executes on the platform that compiles it. A leader-only-kill mutation turns it RED (`elapsed=10.003s` against the `<4s` assert), GREEN with the teardown restored.

One honest limit: CI runs tests on ubuntu only, so the windows branch is compile-verified (the test file type-checks verbatim against `x86_64-pc-windows-msvc`), not executed. If you want a windows test job added to pr-checks, happy to do that as a follow-up.

Also folded the receive-pack POST block into `post_pack_round` (CodeRabbit's round-1 nitpick); behavior identical, receive-pack unit tests green.

**Review by jatmn — CHANGES_REQUESTED** [review]
I found issues that need to be addressed before this is ready.

## Findings

- [P2] Count a multi-round negotiation as one fetch completion
  `crates/git-remote-gitlawb/src/main.rs:604`
  Each flush-terminated have batch now produces a successful `git-upload-pack` POST, but the server increments `gitlawb_fetches_total` after every successful POST even though that metric is documented as completed fetches. An incremental fetch with N negotiation rounds is therefore reported as N completed fetches (and the repeatedly accumulated request bodies are observed as separate pack sizes). Update the protocol/metrics boundary so intermediate stateless-RPC rounds are not counted as completed fetches, or redefine and separately expose request-level metrics.

- [P2] Keep the timeout in force after the fetch leader exits
  `crates/git-remote-gitlawb/tests/real_git_fetch.rs:393`
  `try_wait()` marks the run complete as soon as the top-level `git` process exits, then the code joins the stdout/stderr drainers without any deadline. A remote-helper descendant can still hold those inherited pipe write ends after its parent exits, making either join wait for the helper's 300-second HTTP timeout (or indefinitely) despite the requested bound. The existing process-group cleanup only runs when the leader is still alive at the deadline, so add coverage for a leader-exits-first descendant and keep/reap the process group until the pipe readers have completed within the deadline.

**Comment by beardthelion** [comment]
Both addressed on `e5baece` (F1) and `6e78eb2` (F2), each verified by execution.

**[P2] Count a multi-round negotiation as one fetch completion.** `git_upload_pack` now counts a completed fetch (and observes the pack) only on the POST that completes one: the finalizing `done` round on the plain path, or the single POST that reaches the filtered path (`upload_pack_excluding` serves a self-contained pack with no `done`, so gating purely on `done` would have dropped path-scoped fetches to zero). A completed fetch is exactly one such POST. Verified: a three-round plain fetch and a no-`done` filtered serve each record one completion; counting per POST records 3 for the plain fetch. (`observe_pack_size` still measures the request body, not the served pack; that mislabel predates this PR and I've left it as a separate follow-up.)

**[P2] Keep the timeout in force after the fetch leader exits.** `run_bounded` now closes the process group on the clean-exit path too, not only at the deadline. When the git leader exits but a descendant still holds the inherited pipe write-ends, the reader joins would otherwise block on the helper's ~300s timeout; the group is signalled behind a `kill(-pgid, 0)` liveness probe so a reaped-and-recycled pid is never hit. Added a leader-exits-first fixture and regression: with the clean-path reap removed it blocks ~10s on the descendant, with it the joins return promptly. Unix-only, matching the existing teardown.

**Review by jatmn — CHANGES_REQUESTED** [review]
I found issues that need to be addressed before this is ready.

## Findings

- [P2] Count successful `no-done` fetches
  `crates/gitlawb-node/src/api/repos.rs:709`
  The node advertises `multi_ack_detailed no-done`, so a normal smart-HTTP client may finish a flush-terminated have round by receiving `ACK <oid> ready` followed immediately by the pack, without ever sending a `done` pkt-line. I reproduced that response with this checkout's exact `git upload-pack --stateless-rpc` path. In that successful case both `finalizes_fetch` and `served_filtered_pack` are false, so this gate records neither `gitlawb_fetches_total` nor the pack observation. Please derive completion from the response/protocol outcome rather than assuming every plain fetch ends with `done`, and add a no-`done` regression.

- [P2] Do not count rejected filtered responses as completed fetches
  `crates/gitlawb-node/src/api/repos.rs:686`
  Setting `served_filtered_pack` for every `upload_pack_excluding` response labels the request as a completed fetch even when the client rejects the response. The committed real-Git scenario currently sends two POSTs and exits nonzero with `expected ACK/NAK`; against this handler both filtered POSTs would set this flag and increment the completion counter, reporting one failed fetch as two successful completions. Until #191 makes this negotiation valid, gate the metric on a genuinely final request/outcome (fresh filtered clones carry `done`) or use a request-level metric instead.

- [P2] Keep the clean-exit timeout bounded on Windows
  `crates/git-remote-gitlawb/tests/real_git_fetch.rs:431`
  The new leader-exits-first cleanup runs only under `cfg(unix)`. On Windows, if the top-level Git process exits while its remote-helper descendant remains blocked in the 300-second HTTP request, `completed` becomes true and execution goes straight to the reader-thread joins while that descendant still owns the pipe handles. The 30-second runner can therefore block for minutes; the matching regression is also Unix-only, and `taskkill /T` cannot recover the tree after the leader has exited. Use a Windows Job Object or another ownership mechanism established at spawn time, and exercise this ordering on Windows.

- [P2] Avoid racing the process-wide metrics initializer in tests
  `crates/gitlawb-node/src/api/repos.rs:1937`
  This adds a second test that calls `metrics::init`, while the existing metrics test does the same and `init` uses an unsynchronized `REGISTRY.get()` check followed by several `OnceLock::set(...).expect(...)` calls. If both tests observe an empty registry, the loser panics at `set INFO once` (or a later lock); selecting these two tests with two test threads reproduced that panic immediately. Use an atomic one-time initializer such as `get_or_init`/`Once`, or make this test assert the pure decision without invoking the global initializer.

- [P2] Reject malformed pkt-line headers instead of treating them as flushes
  `crates/git-remote-gitlawb/src/main.rs:450`
  Both UTF-8 and hex parse failures are converted to zero, so a header such as `zzzz` is handled exactly like the valid `0000` flush. An invalid opener followed by EOF now returns success without a POST, and an invalid terminator after valid have lines can make those lines go out as a synthesized, potentially signed flush round. This masks/desynchronizes corrupted protocol input; propagate either parse error and add non-hex/non-UTF-8 cases alongside the existing `0001` test.

- [P2] Bind the withheld-path test to the expected rejection
  `crates/git-remote-gitlawb/tests/real_git_fetch.rs:736`
  Once one POST has reached the shim, this branch accepts every nonzero Git exit as the intended withheld-pack rejection. A truncated/corrupt response, later HTTP failure, or helper protocol regression after the first POST therefore leaves the test green even though the comments claim it proved the helper forwarded and terminated correctly. The current runtime error is specifically `expected ACK/NAK`; assert that failure signature (and the expected request/response shape) so unrelated post-POST failures cannot masquerade as the recorded #191 limitation.

**Comment by beardthelion** [comment]
Pushed `3ceba33`. Each fix has an adversarial regression that fails on the old behavior and passes on the new; the full workspace suite and clippy are green.

- **Count `no-done` fetches / don't count rejected filtered (`repos.rs`):** the completion metric keys off the response outcome now, not the request. `upload_pack` / `upload_pack_excluding` report whether they served a pack (`response_served_pack` finds the PACK magic in both side-band-64k and raw framing). The plain path counts when a pack was served, so a `no-done` `ACK <oid> ready` + pack counts once and a negotiation-only round counts zero; the filtered path stays gated on the finalizing `done` round, so the pre-#191 two-POST rejection no longer double-counts. `no_done_plain_fetch_counts_once` and `rejected_filtered_fetch_counts_zero_not_two` are red on the old `finalizes || served_filtered` rule.
- **Reject malformed pkt-line headers (`main.rs`):** a non-UTF-8 or non-hex length prefix errors now instead of collapsing to a `0000` flush; a real `0000` still parses as a flush. Added non-hex and non-UTF-8 cases.
- **Metrics init race:** `metrics::init` is guarded by `std::sync::Once`, so concurrent callers can't both pass the old `REGISTRY.get()` check and race the `OnceLock` setters. A 32-thread barrier stress-test panics before the fix and passes after.
- **Bind the withheld-path test:** it asserts the specific `expected ACK/NAK` signature now (with `LC_ALL=C` pinned so a localized git build doesn't translate it away), so a truncated response or a later failure after the first POST no longer passes as the withheld rejection.
- **`request_finalizes_fetch` framing:** rejects `0003` framing and trailing bytes after `done`; `00030009done\n` and `0009done\nx` return false.

On the Windows clean-exit timeout: I'm leaving the Windows ownership mechanism for a follow-up rather than shipping it here. CI is Linux-only (every `pr-checks` job is `ubuntu-latest`) and the matching regression is `cfg(unix)`, so a Windows Job Object implementation would go in untested, which I'd rather not do in this PR.

Re-requesting your review.

**Review by jatmn — CHANGES_REQUESTED** [review]
I found issues that need to be addressed before this is ready.

## Findings

- [P2] Keep the clean-exit harness bounded on Windows
  `crates/git-remote-gitlawb/tests/real_git_fetch.rs:435`
  The clean-exit cleanup that prevents a remote-helper descendant from retaining the reader pipes is compiled only on Unix. On Windows, `try_wait` can observe the `git fetch` leader exiting while `git-remote-gitlawb` remains blocked in its 300-second HTTP request with inherited stdout/stderr handles; the following reader joins then wait for that descendant, despite the runner's 30-second bound. `taskkill /T` is only usable while the leader remains alive, and the corresponding regression is Unix-only. Establish tree ownership at spawn time (for example, with a Job Object) and tear it down before joining, with a Windows regression.

**Comment by beardthelion** [comment]
Implemented, reversing my earlier deferral on this. Pushed as `443908d`.

Last round I punted the Windows clean-exit case as a follow-up on the grounds that CI is Linux-only. That was the wrong call, and you were right to keep pushing: `git-remote-gitlawb` is a *shipped* Windows binary (`release.yml` builds `x86_64-pc-windows-msvc` and ships `gl` + `git-remote-gitlawb`), so deferring it meant shipping the untested path, not avoiding it.

`run_bounded` now assigns the child to a **Job Object** at spawn (the analog of the Unix `process_group(0)`) and terminates the job on both the timeout and clean-exit paths, so the whole fetch tree dies regardless of leader liveness. That closes exactly the hole you described: the clean-exit reap was `#[cfg(unix)]`-only, and `taskkill /T` can't reach a descendant once the leader has exited, whereas terminating the job doesn't depend on the leader. `KILL_ON_JOB_CLOSE` is the safety net for a stray. The clean-exit regression is now cross-platform so it exercises this teardown.

On verification, honestly: the Unix path is unchanged and green (44 + 6 tests, clippy/fmt clean). The Windows path is compile-verified for `x86_64-pc-windows-msvc` — the full-crate cross-check is blocked by `ring`'s C build not cross-compiling on the aarch64 dev host, so I type-checked the complete `run_bounded` + Job Object code in isolation against the msvc target instead. Its runtime behavior is unverified until it actually runs on Windows, which is the real gap underneath this: we ship the binary but never run its tests on Windows. I filed that as #228 (a non-blocking Windows CI lane), which is what makes this regression load-bearing rather than never-run.

**Review by jatmn — CHANGES_REQUESTED** [review]
I found issues that need to be addressed before this is ready.

## Findings

- [P2] Keep the clean-exit harness bounded when Job Object setup fails
  `crates/git-remote-gitlawb/tests/real_git_fetch.rs:579`
  `assign_job` deliberately returns `None` when creating, configuring, or assigning the Job Object fails, but this clean-exit branch only tears down `Some(job)` before joining the two pipe drainers. In a Windows environment where assignment is unavailable (for example, an inherited/restricted job), `git fetch` can exit while its remote-helper descendant still owns the inherited pipe handles; the joins then wait for the helper's 300-second HTTP timeout. The timeout branch has a fallback, but this is precisely the leader-exits-first path that fallback cannot recover after the parent is gone. Keep this path bounded when no Job Object was established (or make establishment mandatory and fail the test explicitly).

**Comment by beardthelion** [comment]
Addressed in 050536a. I made the Windows Job Object mandatory rather than trying to bound the clean-exit path without it: as you noted, taskkill `/T` can't reach the orphaned helper once the leader has exited, and std threads have no join-with-timeout, so the job is the only reliable teardown here. `assign_job` now returns `JobHandle` and panics per-step if the job can't be established, and both teardown paths (timeout and clean-exit) unconditionally terminate it. That closes the clean-exit gap and turns a silent ~300s hang into a loud setup failure; `KILL_ON_JOB_CLOSE` also kills the descendant at latest when the handle drops. `reap_tree`, whose only Windows use was the removed `None` arm, is now `cfg(all(not(unix), not(windows)))`.

Verification, split honestly since this is all `cfg(windows)`: I cross-compiled the tests for `x86_64-pc-windows-gnu` (compiles clean, `clippy -D warnings` clean, no dead_code from the reap_tree change) and confirmed the Linux `git-remote-gitlawb` suite stays green, including the unix analog of this scenario. The actual Windows runtime teardown I did not execute locally, so CI's Windows job is the gate for that.

@jatmn ready for another look.

**Review by jatmn — APPROVED** [review]
@beardthelion LGTM

**Comment by beardthelion** [comment]
@kevincodex1 this one's ready to merge. jatmn approved the current head (`050536a`) and CI is green across the board. It's the multi-round fetch fix for #117, open since the 11th and the oldest thing in my queue.

**Review by jatmn — APPROVED** [review]
LGTM

**Comment by kevincodex1** [comment]
hello bro @beardthelion please rebase to main and fix conflicts

**Review by kevincodex1 — APPROVED** [review]

**Comment by beardthelion** [comment]
@kevincodex1 rebased onto main and resolved, now at a53c151.

The conflicts were in `api/repos.rs` and `git/smart_http.rs`. #174 had changed `upload_pack` and `upload_pack_excluding` to thread the admission guard so permits release only after the git process group is reaped, and this branch had changed the same two to return the served-pack flag the fetch metric keys off. Both are kept.

CI is green. One note for the history: the Windows lane failed once on a localhost connection abort in `real_git_withheld_shaped_first_post`, before the test's shim saw a POST, and passed on re-run. Nothing that lane compiles changed in the rebase apart from the 0.7.1 version bump, so that one is a flake in the test rather than anything in the diff. Happy to file it separately if you want it fixed before this goes in.

**Comment by beardthelion** [comment]
Correcting my comment above. I called that Windows failure a flake and offered to file it separately. It is a real bug in the test shim, this PR introduced it, and it is on main now. The fix is #312.

`start_shim` makes the listener non-blocking so its accept loop can poll the stop flag. On unix `accept()` hands back a blocking socket, so it never showed here. On Windows the accepted socket inherits the listener's FIONBIO state and std does not reset it, so every connection arrives non-blocking. `handle_conn` then reads with `read_line(..).unwrap_or(0)`, which turns the resulting `WouldBlock` into "peer closed" and returns without writing a response, so the connection is dropped and the client aborted mid-request. That is the `os error 10053`, and the zero POSTs, because the shim returns before the branch that counts them.

The window is open on every connection, since the shim accepts as soon as the handshake completes and that is strictly before the request bytes arrive. It survives only because the 5ms accept poll usually lags the client, which is what made it look intermittent. Both shim-based tests are exposed, not just the one that lost the race.

I should have run this down before saying CI was green rather than after, since that comment is what this merged on.
